### PR TITLE
discovery: fix walk for keys and add debugging

### LINKED ISSUE #166: discovery: don't require ACI endpoints when discovery pubkeys
Discovery currently returns an error (and hence fails completely) in all cases in which it cannot find ACI endpoints:  https://github.com/appc/spec/blob/2fe56d3eeef90f9655e53bbb62ec00d9419a8c27/discovery/discovery.go#L146

This needs to be refactored a bit, since the "discover trusted key" use case does not work with this logic (it's quite possible for there to be no ACI corresponding exactly to the key being trusted, e.g. on coreos.com)

### PR DESCRIPTION
**Pull request by philips**
Add two new functions to let a user control the Discover walking process and
then also add attempts list to keep track of errors that were encountered along
the way.

Fixes #166

### CODE REVIEW

File: discovery/discovery.go [hunk 1]
```diff
@@ -143,26 +143,72 @@ func doDiscover(app App, pre string, insecure bool) (*Endpoints, error) {
...
+// and user feedback.
+type FailedAttempt struct {
+	Prefix string
+	Error  error
+}
+
+// DiscoverEndpoints will make HTTPS requests to find the ac-discovery meta
+// tags and optionally will use HTTP if insecure is set. It will not give up
+// until it has exhausted the path or found an image discovery.
+func DiscoverEndpoints(app App, insecure bool) (eps *Endpoints, attempts []FailedAttempt, err error) {
```
Comment (eyakubovich) [inline discovery/discovery.go]: Wouldn't this cause the whole thing to abort if the first iteration was not fruitful?

File: discovery/discovery.go [hunk 2]
```diff
@@ -143,26 +143,72 @@ func doDiscover(app App, pre string, insecure bool) (*Endpoints, error) {
...
+	dfn := func(pre string, eps *Endpoints, err error) error {
+		if err != nil {
+			attempts = append(attempts, FailedAttempt{pre, err})
+			return nil
+		}
+		if len(eps.ACIEndpoints) == 0 {
+			return fmt.Errorf("found no ACI meta tags")
```
Comment (eyakubovich) [inline discovery/discovery.go]: Wouldn't this cause the whole thing to abort if the first iteration was not fruitful?

Comment (philips) [inline discovery/discovery.go]: Yea, this is how the old code worked too. So, it would assume that the first 200 website it got should have meta info.

**Comment by jonboulle** [comment]
This could do with some cleaning up, but lgtm for now.

**Pull request merged**